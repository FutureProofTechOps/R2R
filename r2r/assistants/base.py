import json
from abc import ABCMeta, abstractmethod
from typing import Any, Optional

from r2r.base import (
    Assistant,
    AsyncSyncMeta,
    GenerationConfig,
    Message,
    syncable,
)


class CombinedMeta(AsyncSyncMeta, ABCMeta):
    pass


class BaseR2RAssistant(Assistant, metaclass=CombinedMeta):

    @syncable
    async def arun(
        self,
        system_instruction: Optional[str] = None,
        messages: Optional[list[Message]] = None,
    ) -> str:
        messages_have_system_instruction = not (
            messages and "system" in messages
        )
        if system_instruction or messages_have_system_instruction:
            self._completed = False
            self._setup(system_instruction)

        if messages:
            for message in messages:
                print("message = ", message)
                self.conversation.add_message(message)

        while not self._completed:
            generation_config_with_functions = self.get_generation_config()
            result = await self.get_and_process_response(
                generation_config_with_functions
            )

            if self._completed:
                return result

    def get_generation_config(self) -> GenerationConfig:
        return GenerationConfig(
            **self.config.generation_config.model_dump(
                exclude={"functions", "stream"},
            ),
            functions=[
                {
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": tool.parameters,
                }
                for tool in self.tools
            ],
            stream=self.is_streaming(),
        )

    @abstractmethod
    async def get_and_process_response(
        self, generation_config: GenerationConfig
    ) -> str:
        pass

    @abstractmethod
    def is_streaming(self) -> bool:
        pass

    async def handle_function_call(
        self, function_name: str, function_arguments: str
    ) -> str:
        tool_args = json.loads(function_arguments)
        self.conversation.create_and_add_message(
            "assistant",
            function_call={
                "name": function_name,
                "arguments": function_arguments,
            },
        )
        tool_result = await self.execute_tool(function_name, **tool_args)
        self.conversation.create_and_add_message(
            "function", content=tool_result, name=function_name
        )
        return await self.arun()  # Continue the conversation

    def handle_content_response(self, content: str) -> str:
        self.conversation.create_and_add_message("assistant", content=content)
        self._completed = True
        return content


class R2RAssistant(BaseR2RAssistant):
    def is_streaming(self) -> bool:
        return False

    async def get_and_process_response(
        self, generation_config: GenerationConfig
    ) -> str:
        response = await self.llm_provider.aget_completion(
            self.conversation.get_messages(),
            generation_config,
        )
        return await self.process_llm_response(response)

    async def process_llm_response(self, response: dict[str, Any]) -> str:
        message = response.choices[0].message
        if message.function_call:
            return await self.handle_function_call(
                message.function_call.name, message.function_call.arguments
            )
        else:
            return self.handle_content_response(message.content)


class R2RStreamingAssistant(BaseR2RAssistant):
    def is_streaming(self) -> bool:
        return True

    async def get_and_process_response(
        self, generation_config: GenerationConfig
    ) -> str:
        stream = self.llm_provider.get_completion_stream(
            self.conversation.get_messages(),
            generation_config,
        )
        return await self.process_llm_response(stream)

    async def process_llm_response(self, stream):
        function_name = None
        function_arguments = ""
        content = ""

        for chunk in stream:
            print("chunk = ", chunk)
            delta = chunk.choices[0].delta

            if delta.function_call:
                if delta.function_call.name:
                    function_name = delta.function_call.name
                if delta.function_call.arguments:
                    function_arguments += delta.function_call.arguments
            elif delta.content:
                content += delta.content

        if function_name:
            return await self.handle_function_call(
                function_name, function_arguments
            )
        else:
            return self.handle_content_response(content)
