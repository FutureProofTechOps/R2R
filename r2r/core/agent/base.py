"""Defines the abstract base classes and enums for agent objects"""

import logging
import logging.config
from abc import ABC, abstractmethod
from typing import Sequence

from ..abstractions.output import LLMChatCompletion
from ..abstractions.tool import Tool
from ..providers.llm import LLMProvider
from ..providers.prompt import PromptProvider

logger = logging.getLogger(__name__)


class Agent(ABC):
    """
    An abstract class for implementing a agent.

    An agent is an autonomous entity that can perform actions and communicate
    with other providers.
    """

    def __init__(
        self, prompt_provider: PromptProvider, llm_provider: LLMProvider
    ) -> None:
        self.llm_provider = llm_provider
        self.prompt_provider = prompt_provider

        self._initialized = False

    @abstractmethod
    def __iter__(self):
        pass

    @abstractmethod
    def __next__(self) -> LLMChatCompletion:
        """
        Iterates the agent by performing a single step of its task.

        A single step is a new conversation turn, which consists of generating
        a new 'asisstant' message, and parsing the reply from the 'user'.

        Raises:
            AgentStopIterationError: If the agent has already completed its task
            or exceeded the maximum number of iterations.
        """
        pass

    @property
    @abstractmethod
    def tools(self) -> Sequence[Tool]:
        """An abstract property for getting the tools associated with the agent."""
        pass

    @abstractmethod
    def run(self, message: str) -> str:
        """
        Runs the agent until it completes its task.

        The task is complete when next returns None.
        """
        pass

    @abstractmethod
    def run_stream(self, message: str):
        """
        Runs the agent until it completes its task. Stream the result when it is ready.
        """
        pass

    @abstractmethod
    def _setup(self) -> None:
        """An abstract method for setting up the agent before running."""
        pass
