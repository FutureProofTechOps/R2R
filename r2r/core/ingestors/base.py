import json
import string
from abc import ABC, abstractmethod
from io import BytesIO
from typing import Any, Generator, Generic, TypeVar, Union

from bs4 import BeautifulSoup

T = TypeVar("T")


class Ingestor(ABC, Generic[T]):
    @abstractmethod
    def ingest(self, data: Any) -> Generator[str, None, None]:
        pass


class TextAdapter(Ingestor[str]):
    def ingest(self, data: str) -> Generator[str, None, None]:
        yield data


class JSONAdapter(Ingestor[str]):
    def ingest(self, data: Union[str, bytes]) -> Generator[str, None, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        yield self._parse_json(json.loads(data))

    def _parse_json(self, data: dict) -> str:
        def remove_objects_with_null(obj):
            if not isinstance(obj, dict):
                return obj
            result = obj.copy()
            for key, value in obj.items():
                if isinstance(value, dict):
                    result[key] = remove_objects_with_null(value)
                elif value is None:
                    del result[key]
            return result

        def format_json_as_text(obj, indent=0):
            lines = []
            indent_str = " " * indent

            if isinstance(obj, dict):
                for key, value in obj.items():
                    if isinstance(value, (dict, list)):
                        nested = format_json_as_text(value, indent + 2)
                        lines.append(f"{indent_str}{key}:\n{nested}")
                    else:
                        lines.append(f"{indent_str}{key}: {value}")
            elif isinstance(obj, list):
                for item in obj:
                    nested = format_json_as_text(item, indent + 2)
                    lines.append(f"{nested}")
            else:
                return f"{indent_str}{obj}"

            return "\n".join(lines)

        return format_json_as_text(remove_objects_with_null(data))


class HTMLAdapter(Ingestor[str]):
    def ingest(self, data: str) -> list[str]:
        soup = BeautifulSoup(data, "html.parser")
        yield soup.get_text()


class PDFAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from pypdf import PdfReader

            self.PdfReader = PdfReader
        except ImportError:
            raise ValueError(
                "Error, `pypdf` is required to run `PyPDFAdapter`. Please install it using `pip install pypdf`."
            )

    def ingest(self, data: bytes) -> Generator[str, None, None]:
        pdf = self.PdfReader(BytesIO(data))
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text is not None:
                page_text = "".join(
                    filter(lambda x: x in string.printable, page_text)
                )
                yield page_text


class PPTAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from pptx import Presentation

            self.Presentation = Presentation
        except ImportError:
            raise ValueError(
                "Error, `python-pptx` is required to run `PPTAdapter`. Please install it using `pip install python-pptx`."
            )

    def ingest(self, data: bytes) -> list[str]:
        prs = self.Presentation(BytesIO(data))
        for slide in prs.slides:
            for shape in slide.shapes:
                if hasattr(shape, "text"):
                    yield shape.text


class DOCXAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from docx import Document

            self.Document = Document
        except ImportError:
            raise ValueError(
                "Error, `python-docx` is required to run `DOCXAdapter`. Please install it using `pip install python-docx`."
            )

    def ingest(self, data: bytes) -> Generator[str, None, None]:
        doc = self.Document(BytesIO(data))
        for paragraph in doc.paragraphs:
            yield paragraph.text


class CSVAdapter(Ingestor[str]):
    def __init__(self):
        import csv
        from io import StringIO

        self.csv = csv
        self.StringIO = StringIO

    def ingest(self, data: Union[str, bytes]) -> Generator[str, None, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        csv_reader = self.csv.reader(self.StringIO(data))
        for row in csv_reader:
            yield ", ".join(row)


class XLSXAdapter(Ingestor[str]):
    def __init__(self):
        try:
            from openpyxl import load_workbook

            self.load_workbook = load_workbook
        except ImportError:
            raise ValueError(
                "Error, `openpyxl` is required to run `XLSXAdapter`. Please install it using `pip install openpyxl`."
            )

    def ingest(self, data: bytes) -> Generator[str, None, None]:
        wb = self.load_workbook(filename=BytesIO(data))
        for sheet in wb.worksheets:
            for row in sheet.iter_rows(values_only=True):
                yield ", ".join(map(str, row))


class MarkdownAdapter(Ingestor[str]):
    def __init__(self):
        import markdown

        self.markdown = markdown

    def ingest(self, data: str) -> Generator[str, None, None]:
        html = self.markdown.markdown(data)
        soup = BeautifulSoup(html, "html.parser")
        yield soup.get_text()
