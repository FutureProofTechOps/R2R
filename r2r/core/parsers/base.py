import json
import string
from abc import ABC, abstractmethod
from io import BytesIO
from typing import Any, AsyncGenerator, Generic, TypeVar, Union

from bs4 import BeautifulSoup

from ..abstractions.document import DataType

T = TypeVar("T")


class Parser(ABC, Generic[T]):
    @abstractmethod
    async def ingest(self, data: Any) -> AsyncGenerator[DataType, None]:
        pass


class TextParser(Parser[str]):
    async def ingest(self, data: str) -> AsyncGenerator[DataType, None]:
        yield data


class JSONParser(Parser[str]):
    async def ingest(
        self, data: Union[str, bytes]
    ) -> AsyncGenerator[str, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        yield self._parse_json(json.loads(data))

    def _parse_json(self, data: dict) -> str:
        def remove_objects_with_null(obj):
            if not isinstance(obj, dict):
                return obj
            result = obj.copy()
            for key, value in obj.items():
                if isinstance(value, dict):
                    result[key] = remove_objects_with_null(value)
                elif value is None:
                    del result[key]
            return result

        def format_json_as_text(obj, indent=0):
            lines = []
            indent_str = " " * indent

            if isinstance(obj, dict):
                for key, value in obj.items():
                    if isinstance(value, (dict, list)):
                        nested = format_json_as_text(value, indent + 2)
                        lines.append(f"{indent_str}{key}:\n{nested}")
                    else:
                        lines.append(f"{indent_str}{key}: {value}")
            elif isinstance(obj, list):
                for item in obj:
                    nested = format_json_as_text(item, indent + 2)
                    lines.append(f"{nested}")
            else:
                return f"{indent_str}{obj}"

            return "\n".join(lines)

        return format_json_as_text(remove_objects_with_null(data))


class HTMLParser(Parser[str]):
    async def ingest(self, data: str) -> AsyncGenerator[str, None]:
        soup = BeautifulSoup(data, "html.parser")
        yield soup.get_text()


class PDFParser(Parser[str]):
    def __init__(self):
        try:
            from pypdf import PdfReader

            self.PdfReader = PdfReader
        except ImportError:
            raise ValueError(
                "Error, `pypdf` is required to run `PyPDFParser`. Please install it using `pip install pypdf`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        pdf = self.PdfReader(BytesIO(data))
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text is not None:
                page_text = "".join(
                    filter(lambda x: x in string.printable, page_text)
                )
                yield page_text


class PPTParser(Parser[str]):
    def __init__(self):
        try:
            from pptx import Presentation

            self.Presentation = Presentation
        except ImportError:
            raise ValueError(
                "Error, `python-pptx` is required to run `PPTParser`. Please install it using `pip install python-pptx`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        prs = self.Presentation(BytesIO(data))
        for slide in prs.slides:
            for shape in slide.shapes:
                if hasattr(shape, "text"):
                    yield shape.text


class DOCXParser(Parser[str]):
    def __init__(self):
        try:
            from docx import Document

            self.Document = Document
        except ImportError:
            raise ValueError(
                "Error, `python-docx` is required to run `DOCXParser`. Please install it using `pip install python-docx`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        doc = self.Document(BytesIO(data))
        for paragraph in doc.paragraphs:
            yield paragraph.text


class CSVParser(Parser[str]):
    def __init__(self):
        import csv
        from io import StringIO

        self.csv = csv
        self.StringIO = StringIO

    async def ingest(
        self, data: Union[str, bytes]
    ) -> AsyncGenerator[str, None]:
        if isinstance(data, bytes):
            data = data.decode("utf-8")
        csv_reader = self.csv.reader(self.StringIO(data))
        for row in csv_reader:
            yield ", ".join(row)


class XLSXParser(Parser[str]):
    def __init__(self):
        try:
            from openpyxl import load_workbook

            self.load_workbook = load_workbook
        except ImportError:
            raise ValueError(
                "Error, `openpyxl` is required to run `XLSXParser`. Please install it using `pip install openpyxl`."
            )

    async def ingest(self, data: bytes) -> AsyncGenerator[str, None]:
        wb = self.load_workbook(filename=BytesIO(data))
        for sheet in wb.worksheets:
            for row in sheet.iter_rows(values_only=True):
                yield ", ".join(map(str, row))


class MarkdownParser(Parser[str]):
    def __init__(self):
        import markdown

        self.markdown = markdown

    async def ingest(self, data: str) -> AsyncGenerator[str, None]:
        html = self.markdown.markdown(data)
        soup = BeautifulSoup(html, "html.parser")
        yield soup.get_text()
