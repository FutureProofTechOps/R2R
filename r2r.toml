[auth]
provider = "r2r"
access_token_lifetime_in_minutes = 60
refresh_token_lifetime_in_days = 7
require_authentication = false
require_email_verification = false
default_admin_email = "admin@example.com"
default_admin_password = "change_me_immediately"

[completion]
provider = "litellm"
concurrent_request_limit = 16

  [completion.generation_config]
  model = "openai/gpt-4o"
  temperature = 0.1
  top_p = 1
  max_tokens_to_sample = 1_024
  stream = false
  add_generation_kwargs = { }

[crypto]
provider = "bcrypt"

[database]
provider = "postgres"

[embedding]
provider = "litellm"
base_model = "text-embedding-3-small"
base_dimension = 512
batch_size = 128
add_title_as_prefix = false
rerank_model = "None"
concurrent_request_limit = 256

[eval]
provider = "None"

[ingestion]
  [[ingestion.override_parsers]]
  document_type = "pdf"
  parser = "PDFParserUnstructured"

  [ingestion.text_splitter]
  type = "recursive_character"
  chunk_size = 512
  chunk_overlap = 20

[kg]
provider = "None"

[logging]
provider = "local"
log_table = "logs"
log_info_table = "log_info"

[prompt]
provider = "r2r"
