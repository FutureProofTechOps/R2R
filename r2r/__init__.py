from .core import *
from .eval import *
from .main import *
from .pipes import *
from .prompts import *

__all__ = [
    "LoggingConfig",
    "LocalPipeLoggingProvider",
    "PostgresLoggingConfig",
    "PostgresPipeLoggingProvider",
    "RedisLoggingConfig",
    "RedisPipeLoggingProvider",
    "PipeLoggingConnectionSingleton",
    "VectorEntry",
    "VectorType",
    "Vector",
    "SearchRequest",
    "SearchResult",
    "AsyncPipe",
    "PipeRunInfo",
    "PipeType",
    "AsyncState",
    "Prompt",
    "DataType",
    "DocumentType",
    "Document",
    "Extraction",
    "ExtractionType",
    "Fragment",
    "FragmentType",
    # Parsers
    "AsyncParser",
    "CSVParser",
    "DOCXParser",
    "HTMLParser",
    "JSONParser",
    "MarkdownParser",
    "PDFParser",
    "PPTParser",
    "TextParser",
    "XLSXParser",
    "Pipeline",
    # Providers
    "EmbeddingConfig",
    "EmbeddingProvider",
    "EvalConfig",
    "EvalProvider",
    "LLMEvalProvider",
    "PromptConfig",
    "PromptProvider",
    "GenerationConfig",
    "LLMChatCompletion",
    "LLMChatCompletionChunk",
    "LLMConfig",
    "LLMProvider",
    "VectorDBConfig",
    "VectorDBProvider",
    "R2RConfig",
    "TextSplitter",
    "RecursiveCharacterTextSplitter",
    "generate_run_id",
    "generate_id_from_label",
    "R2RApp",
    # Pipes
    "DefaultEmbeddingPipe",
    "DefaultEvalPipe",
    "DefaultDocumentParsingPipe",
    "DefaultQueryTransformPipe",
    "DefaultRAGPipe",
    "DefaultStreamingRAGPipe",
    "DefaultVectorSearchPipe",
    "DefaultVectorStoragePipe",
    "DefaultPromptProvider",
]
