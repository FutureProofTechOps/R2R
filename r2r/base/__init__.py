from .abstractions import *
from .agent import *
from .api.models import *
from .logging import *
from .parsers import *
from .pipeline import *
from .pipes import *
from .providers import *
from .utils import *

__all__ = [
    ## ABSTRACTIONS
    # Base abstractions
    "AsyncSyncMeta",
    "syncable",
    # Completion abstractions
    "CompletionRecord",
    "MessageType",
    # Document abstractions
    "DataType",
    "Document",
    "DocumentExtraction",
    "DocumentFragment",
    "DocumentInfo",
    "DocumentStatus",
    "DocumentType",
    # Embedding abstractions
    "EmbeddingPurpose",
    "default_embedding_prefixes",
    # Exception abstractions
    "R2RDocumentProcessingError",
    "R2RException",
    # KG abstractions
    "Entity",
    "KGExtraction",
    "Triple",
    "extract_triples",
    # Llama abstractions
    "VectorStoreQuery",
    # LLM abstractions
    "GenerationConfig",
    "LLMChatCompletion",
    "LLMChatCompletionChunk",
    "RAGCompletion",
    # Prompt abstractions
    "Prompt",
    # Search abstractions
    "AggregateSearchResult",
    "KGSearchResult",
    "KGSearchSettings",
    "VectorSearchResult",
    "VectorSearchSettings",
    # User abstractions
    "Token",
    "TokenData",
    "UserStats",
    # Vector abstractions
    "Vector",
    "VectorEntry",
    "VectorType",
    ## AGENT
    # Agent abstractions
    "Agent",
    "AgentConfig",
    "Conversation",
    "Message",
    "Tool",
    "ToolResult",
    ## API
    # Auth Requests
    "CreateUserRequest",
    "DeleteUserRequest",
    "LoginRequest",
    "LogoutRequest",
    "PasswordChangeRequest",
    "PasswordResetConfirmRequest",
    "PasswordResetRequest",
    "RefreshTokenRequest",
    "UserPutRequest",
    "VerifyEmailRequest",
    # Auth Responses
    "GenericMessageResponse",
    "TokenResponse",
    "UserResponse",
    # Ingestion Requests
    "R2RUpdateFilesRequest",
    "R2RIngestFilesRequest",
    # Management Requests
    "R2RUpdatePromptRequest",
    "R2RDeleteRequest",
    "R2RAnalyticsRequest",
    "R2RUsersOverviewRequest",
    "R2RDocumentsOverviewRequest",
    "R2RDocumentChunksRequest",
    "R2RLogsRequest",
    "R2RPrintRelationshipsRequest",
    "R2RCreateGroupRequest",
    "R2RUpdateGroupRequest",
    "R2RAddUserToGroupRequest",
    "R2RRemoveUserFromGroupRequest",
    "R2RGroupsOverviewRequest",
    "R2RScoreCompletionRequest",
    "R2RAssignDocumentToGroupRequest",
    "R2RRemoveDocumentFromGroupRequest",
    # Retrieval Requests
    "R2RSearchRequest",
    "R2RRAGRequest",
    "R2RAgentRequest",
    ## LOGGING
    # Basic types
    "RunType",
    "AnalysisTypes",
    "LogAnalytics",
    "LogAnalyticsConfig",
    "LogFilterCriteria",
    "LogProcessor",
    # Logging Providers
    "LocalRunLoggingProvider",
    "LoggingConfig",
    "PostgresLoggingConfig",
    "PostgresRunLoggingProvider",
    "RedisLoggingConfig",
    "RedisRunLoggingProvider",
    "RunLoggingSingleton",
    # Run Manager
    "RunManager",
    "manage_run",
    ## PARSERS
    # Base parser
    "AsyncParser",
    ## PIPELINE
    # Base pipeline
    "AsyncPipeline",
    ## PIPES
    "AsyncPipe",
    "AsyncState",
    "PipeType",
    ## PROVIDERS
    # Base provider classes
    "Provider",
    "ProviderConfig",
    # Auth provider
    "AuthConfig",
    "AuthProvider",
    # Chunking provider
    "ChunkingConfig",
    "ChunkingProvider",
    "Method",
    # Crypto provider
    "CryptoConfig",
    "CryptoProvider",
    # Database providers
    "DatabaseConfig",
    "DatabaseProvider",
    "RelationalDBProvider",
    "VectorDBProvider",
    "VectorDBFilterValue",
    # Embedding provider
    "EmbeddingConfig",
    "EmbeddingProvider",
    # Knowledge Graph provider
    "KGConfig",
    "KGDBProvider",
    "update_kg_prompt",
    "extract_entities",
    # LLM provider
    "CompletionConfig",
    "CompletionProvider",
    # Parsing provider
    "ParsingConfig",
    "ParsingProvider",
    "OverrideParser",
    # Prompt provider
    "PromptConfig",
    "PromptProvider",
    ## UTILS
    "RecursiveCharacterTextSplitter",
    "TextSplitter",
    "run_pipeline",
    "to_async_generator",
    "generate_run_id",
    "generate_id_from_label",
    "increment_version",
    "EntityType",
    "Relation",
    "format_entity_types",
    "format_relations",
]
