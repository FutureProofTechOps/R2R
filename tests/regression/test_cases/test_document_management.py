import os

from r2r import R2RException
from tests.regression.test_cases.base import BaseTest


class TestDocumentManagement(BaseTest):

    CHUNKS_FILE_ID = "b736292c-11e6-5453-9686-055da3edb866"
    UPDATE_FILE_ID = "93123a68-d668-51de-8291-92162730dc87"
    DELETE_FILE_ID = "b736292c-11e6-5453-9686-055da3edb866"

    def __init__(self, client):
        super().__init__(client)
        self.set_exclude_paths(
            "documents_overview",
            [
                f"root['results'][%i]['{field}']"
                for field in ["created_at", "updated_at"]
                for i in range(20)
            ],
        )

    def get_test_cases(self):
        return {
            "ingest_sample_files": lambda client: self.ingest_sample_files_test(
                client
            ),
            "reingest_sample_file": lambda client: self.ingest_sample_files_test(
                client
            ),
            "documents_overview": lambda client: self.documents_overview_test(
                client
            ),
            "document_chunks_test": lambda client: self.document_chunks_test(
                client
            ),
            "update_document": lambda client: self.update_document_test(
                client
            ),
            "delete_document": lambda client: self.delete_document_test(
                client
            ),
            "rerun_documents_overview": lambda client: self.documents_overview_test(
                client
            ),
            "rerun_document_chunks_test": lambda client: self.document_chunks_test(
                client
            ),
        }

    def ingest_sample_files_test(self, client):
        file_path = os.path.abspath(__file__)
        data_path = os.path.join(
            os.path.dirname(file_path),
            "..",
            "..",
            "..",
            "r2r",
            "examples",
            "data",
        )
        try:
            return client.ingest_files(
                [
                    os.path.join(data_path, file_name)
                    for file_name in os.listdir(data_path)
                ]
            )
        except R2RException as e:
            return {"results": str(e)}

    def documents_overview_test(self, client):
        try:
            return client.documents_overview()
        except R2RException as e:
            return {"results": str(e)}

    def document_chunks_test(self, client):
        try:
            # Now delete the file
            chunks_response = client.document_chunks(
                TestDocumentManagement.CHUNKS_FILE_ID
            )
            return chunks_response
        except R2RException as e:
            return {"results": str(e)}

    def update_document_test(self, client):
        try:
            # Now update the file
            file_path = os.path.join(
                os.path.dirname(__file__),
                "..",
                "..",
                "..",
                "r2r",
                "examples",
                "data",
                "aristotle_v2.txt",
            )
            update_response = client.update_files(
                [file_path], [TestDocumentManagement.UPDATE_FILE_ID]
            )
            return update_response
        except R2RException as e:
            return {"results": str(e)}

    def delete_document_test(self, client):
        try:
            # Now delete the file
            delete_response = client.delete(
                ["document_id"], [TestDocumentManagement.DELETE_FILE_ID]
            )
            return delete_response
        except R2RException as e:
            return {"results": str(e)}
