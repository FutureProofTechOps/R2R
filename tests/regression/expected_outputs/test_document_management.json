{
    "update_document_test": {
      "results": {
        "processed_documents": [
          {
            "id": "db02076e-989a-59cd-98d5-e24e15a0bd27",
            "group_ids": [],
            "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
            "type": "txt",
            "metadata": {},
            "title": "aristotle_v2.txt",
            "version": "v5",
            "size_in_bytes": 7443,
            "status": "success",
            "created_at": "2024-08-13T17:40:47.624486",
            "updated_at": "2024-08-13T17:40:47.624486"
          }
        ],
        "failed_documents": [],
        "skipped_documents": []
      }
    },
    "delete_document_test": {
      "results": "No entries found for deletion."
    },
    "rerun_documents_overview_test": {
      "results": [
        {
          "id": "716fea3a-826b-5b27-8e59-ffbd1a35455a",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "txt",
          "metadata": {},
          "title": "got.txt",
          "version": "v0",
          "size_in_bytes": 9491,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374613Z",
          "updated_at": "2024-08-13T17:16:07.374613Z"
        },
        {
          "id": "52e12576-090f-59db-91f4-6d4b2e29ae6c",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "mp3",
          "metadata": {},
          "title": "sample.mp3",
          "version": "v0",
          "size_in_bytes": 121671,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374619Z",
          "updated_at": "2024-08-13T17:16:07.374619Z"
        },
        {
          "id": "3e157b3a-8469-51db-90d9-52e7d896b49b",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "pdf",
          "metadata": {},
          "title": "uber_2021.pdf",
          "version": "v0",
          "size_in_bytes": 1880483,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374624Z",
          "updated_at": "2024-08-13T17:16:07.374624Z"
        },
        {
          "id": "01d514a4-5cb1-5c86-be8c-0c5c312a02c9",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "mp3",
          "metadata": {},
          "title": "sample2.mp3",
          "version": "v0",
          "size_in_bytes": 72456,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374632Z",
          "updated_at": "2024-08-13T17:16:07.374632Z"
        },
        {
          "id": "c3291abf-8a4e-5d9d-80fd-232ef6fd8526",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "txt",
          "metadata": {},
          "title": "aristotle_v2.txt",
          "version": "v0",
          "size_in_bytes": 7443,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374638Z",
          "updated_at": "2024-08-13T17:16:07.374638Z"
        },
        {
          "id": "e797da22-8c5d-54e5-bed5-a55954cf6bf9",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "html",
          "metadata": {},
          "title": "pg_essay_4.html",
          "version": "v0",
          "size_in_bytes": 118113,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374643Z",
          "updated_at": "2024-08-13T17:16:07.374643Z"
        },
        {
          "id": "7c1105fc-8f62-5a8f-ac81-fe88f3ec9e4c",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "html",
          "metadata": {},
          "title": "pg_essay_5.html",
          "version": "v0",
          "size_in_bytes": 123779,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374647Z",
          "updated_at": "2024-08-13T17:16:07.374647Z"
        },
        {
          "id": "2f576170-c4f9-5141-a910-a0924f341de4",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "pdf",
          "metadata": {},
          "title": "lyft_2021.pdf",
          "version": "v0",
          "size_in_bytes": 1440303,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374651Z",
          "updated_at": "2024-08-13T17:16:07.374651Z"
        },
        {
          "id": "30f950f0-c692-57c5-b6ec-ff78ccf5ccdc",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "txt",
          "metadata": {},
          "title": "test.txt",
          "version": "v0",
          "size_in_bytes": 20,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374655Z",
          "updated_at": "2024-08-13T17:16:07.374655Z"
        },
        {
          "id": "57eec3df-cf68-5559-a80d-ae3fb55b9af1",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "html",
          "metadata": {},
          "title": "pg_essay_2.html",
          "version": "v0",
          "size_in_bytes": 125112,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374658Z",
          "updated_at": "2024-08-13T17:16:07.374658Z"
        },
        {
          "id": "d421207a-d799-5806-8d67-46b2005b15d4",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "txt",
          "metadata": {},
          "title": "yc_companies.txt",
          "version": "v0",
          "size_in_bytes": 47209,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374670Z",
          "updated_at": "2024-08-13T17:16:07.374670Z"
        },
        {
          "id": "451adbbd-e24e-5c9b-80a8-f424b6c7accb",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "png",
          "metadata": {},
          "title": "screen_shot.png",
          "version": "v0",
          "size_in_bytes": 791764,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374673Z",
          "updated_at": "2024-08-13T17:16:07.374673Z"
        },
        {
          "id": "5b1bd54f-4d70-56b9-a017-a618bc75f94c",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "html",
          "metadata": {},
          "title": "pg_essay_3.html",
          "version": "v0",
          "size_in_bytes": 124915,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374677Z",
          "updated_at": "2024-08-13T17:16:07.374677Z"
        },
        {
          "id": "db02076e-989a-59cd-98d5-e24e15a0bd27",
          "group_ids": [],
          "user_id": "2acb499e-8428-543b-bd85-0d9098718220",
          "type": "txt",
          "metadata": {},
          "title": "aristotle_v2.txt",
          "version": "v5",
          "size_in_bytes": 7443,
          "status": "success",
          "created_at": "2024-08-13T17:16:07.374666Z",
          "updated_at": "2024-08-13T17:40:47.624486Z"
        }
      ]
    },
    "rerun_document_chunks_test": {
      "results": "No chunks found for the given document ID."
    }
  }
